### DISCUSSION TITLE
Trouble import node-mssql into a server action

### DISCUSSION THREAD

**Discussion by KevinGage** [vercel/next.js#63028] [Help]
### Summary

I am using the node-mssql library to query data in server actions.  The form is in a client side component and I am using the useFormState hook.  This actually works fine.  The server action is called and I can fetch data from the database.  However to take full advantage of SQL pools and to keep the code DRY I moved some of the SQL connection logic to its own module, then import it into the server action.  Now when I try to run the project in dev mode I receive compilation errors.  So the structure of the working project is page.jsx (client component), action.js (server action containing all db logic).  The structure of the non-working project is page.jsx (exact same client component), action.js (server action containing logic), db.js (module that I would like to handle the initial connection to the db).  if action.js imports anything from db.js, and if db.js has this line in it anywhere `import sql from "mssql";`, then I receive the following error.
```
Module build failed: UnhandledSchemeError: Reading from "node:stream" is not handled by plugins (Unhandled scheme).
Webpack supports "data:" and "file:" URIs by default.
You may need an additional plugin to handle "node:" URIs.
Import trace for requested module:
node:stream
```
Why is it okay to import mssql in the server action, but not in a module that the server action imports?  Even if it is not imported anywhere else?  I tried marking db.js as "use server" but that did not change the behavior.  I will supply code examples in the "Additional information" box below.

### Additional information

```js
page.jsx

"use client";

import { authenticate } from "@/app/lib/actions";
import { useFormState, useFormStatus } from "react-dom";

export default function Page() {
  const [errorMessage, dispatch] = useFormState(authenticate, undefined);

  return (
    <form action={dispatch}>
      <input type="email" name="email" placeholder="Email" required />
      <input type="password" name="password" placeholder="Password" required />
      <div>{errorMessage && <p>{errorMessage}</p>}</div>
      <LoginButton />
    </form>
  );
}

function LoginButton() {
  const { pending } = useFormStatus();

  return (
    <button aria-disabled={pending} type="submit">
      Login
    </button>
  );
}
```

actions.js (working)
```
"use server";

import { redirect } from "next/navigation";
import sql from "mssql";

export async function authenticate(_currentState, formData) {
  try {
    // LOGIC TO CONNECT TO DATABASE AND VALIDATE LOGIN HERE
  } catch (error) {
    if (error) {
      return "login failed"
    }
  }

  // Login was succesful so redirect to the user page
  redirect("/user");
}

```

actions.js (not working)
```
"use server";

import { redirect } from "next/navigation";
import { getPool } from "@/data/db";

export async function authenticate(_currentState, formData) {
  try {
    const pool = getPool()
    // LOGIC TO QUERY DB AND VALIDATE LOGIN HERE 
  } catch (error) {
    if (error) {
      return "Login failed"
    }
  }

  // Login was succesful so redirect to the user page
  redirect("/user");
}
```

db.js (just the simple fact that I import mssql here causes errors. Tried with "use server" and without)
```
"use server";

import sql from "mssql";

export async function getPool() {
  // Doesnt matter whats in here
}
```
```


### Example

_No response_

**Comment by KevinGage** [comment]
After some more troubleshooting I've realized that this issue only occurs when I am using middleware.  I think my issue is the same as this one https://github.com/vercel/next.js/discussions/59736

**Reply by cweikert** [reply]
Have you found a fix or workaround?

**Reply by KevinGage** [reply]
Yes I found that the problem was actually with the middleware.  If I imported mssql in middleware, or ANY module that middleware imported, then the program wouldn't compile.
I remove mssql from middleware and any modules that it used.  Then everything worked fine again.

**Reply by maxijonson** [reply]
Thanks! I've been having this issue in the middleware as well and didn't think it would happen only there until I found this answer. Unfortunately, I need to use it (FirebaseAdmin) directly inside of the middleware... But that's another issue 😅

**Reply by eokoneyo** [reply]
@KevinGage I had this same issue, turns out the issue happens because middlewares supposedly don't support the node.js runtime, just vercel's edge runtime [see here](https://nextjs.org/docs/pages/building-your-application/routing/middleware#runtime), would have been nice to be able to opt out of this.
